apps/api/src/course/course.resolver.ts

```diff
@@ -28,6 +61,140 @@ export class CourseResolver {
     @Args('filter') filter: FilterInput,
     @Args('courseGroup') courseGroup: CourseGroupInput
   ): Promise<Course[]> {
-    return this.courseService.search(filter, courseGroup)
+    if (filter.periodRange) {
+      const { start, end } = filter.periodRange
+      if (!isTime(start) || !isTime(end)) {
+        throw new BadRequestException({
+          reason: 'INVALID_PERIOD_RANGE',
+          message: 'Start time or end time is invalid',
+        })
+      }
+
+      if (start > end) {
+        throw new BadRequestException({
+          reason: 'INVALID_PERIOD_RANGE',
+          message: 'Start time cannot be later than end time',
+        })
+      }
+    }
+
+    const searchResult = await this.searchService.search<ICourseSearchDocument>({
+      index: this.configService.get<string>('courseIndex'),
+      query: buildCourseQuery({
+        keyword: filter.keyword,
+        genEdTypes: filter.genEdTypes,
+        dayOfWeeks: filter.dayOfWeeks,
+        periodRange: filter.periodRange,
+        studyProgram: courseGroup.studyProgram,
+        semester: courseGroup.semester,
+        academicYear: courseGroup.academicYear,
+      }),
+      from: filter.offset,
+      size: filter.limit,
+    })
+
+    return searchResult.map((item) => item.rawData)
+  }
+}
+
+// build the query
+function buildCourseQuery(filter: ICourseSearchFilter): QueryDslQueryContainer {
```

Comment:
this could be moved into a separate file since it's quite large and constant. `apps/api/src/search/queries/course.ts` or something? along with `ICourseSearchFilter` as well.